engine/src/main/java/net/jqwik/engine/support/types/TypeUsageImpl.java

```diff
@@ -681,18 +641,83 @@ public Optional<TypeUsage> getSuperclass() {
 		if (rawType.getSuperclass() == null) {
 			return Optional.empty();
 		}
-		return Optional.of(TypeUsage.forType(rawType.getGenericSuperclass()));
+		// Double checking for thread safety
+		if (superclass == null) {
+			synchronized (this) {
```

Comment:
This is not safe.
See https://shipilev.net/blog/2014/safe-public-construction/#_safe_publication
`superclass` is not `volatile`, so you must not read the value "without a lock" if you need correctness.
Of course, there are exceptions, however, I am pretty sure this is a plain race condition (unsafe publication) since `TypeUsageImpl` is mutable

Reply:
volatile is definitely necessary.